lib/client/ca_export.go

```diff
@@ -95,35 +104,99 @@ func (r *ExportAuthoritiesRequest) shouldExportIntegration(ctx context.Context)
+}
+
+func exportAllAuthorities(
+	ctx context.Context,
+	client authclient.ClientI,
+	req ExportAuthoritiesRequest,
+	exportSecrets bool,
+) ([]*ExportedAuthority, error) {
+	var authorities []*ExportedAuthority
+	switch isIntegration, err := req.shouldExportIntegration(ctx); {
+	case err != nil:
+		return nil, trace.Wrap(err)
+	case isIntegration && exportSecrets:
+		return nil, trace.NotImplemented("export with secrets is not supported for %q CAs", req.AuthType)
+	case isIntegration:
+		authorities, err = exportAuthForIntegration(ctx, client, req)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	default:
+		authorities, err = exportAuth(ctx, client, req, exportSecrets)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
 	}
-	return exportAuth(ctx, client, req, false /* exportSecrets */)
+
+	// Sanity check that we have at least one authority.
+	// Not expected to happen in practice.
+	if len(authorities) == 0 {
+		return nil, trace.BadParameter("export returned zero authorities")
+	}
+
+	return authorities, nil
+}
+
+// ExportAuthorities is the single-authority version of [ExportAllAuthorities].
+// Soft-deprecated, prefer using [ExportAllAuthorities] and handling exports
+// with more than one authority gracefully.
+func ExportAuthorities(ctx context.Context, client authclient.ClientI, req ExportAuthoritiesRequest) (string, error) {
+	// TODO(codingllama): Remove ExportAuthorities.
+	return exportAuthorities(ctx, client, req, ExportAllAuthorities)
 }
 
-// ExportAuthoritiesSecrets exports the Authority Certificate secrets (private keys).
-// See ExportAuthorities for more information.
+// ExportAuthoritiesSecrets is the single-authority variant of
+// [ExportAllAuthoritiesSecrets].
+// Soft-deprecated, prefer using [ExportAllAuthoritiesSecrets] and handling
+// exports with more than one authority gracefully.
 func ExportAuthoritiesSecrets(ctx context.Context, client authclient.ClientI, req ExportAuthoritiesRequest) (string, error) {
-	if isIntegration, err := req.shouldExportIntegration(ctx); err != nil {
+	// TODO(codingllama): Remove ExportAuthoritiesSecrets.
+	return exportAuthorities(ctx, client, req, ExportAllAuthoritiesSecrets)
+}
+
+func exportAuthorities(
+	ctx context.Context,
+	client authclient.ClientI,
+	req ExportAuthoritiesRequest,
+	exportAllFunc func(context.Context, authclient.ClientI, ExportAuthoritiesRequest) ([]*ExportedAuthority, error),
+) (string, error) {
+	authorities, err := exportAllFunc(ctx, client, req)
+	if err != nil {
 		return "", trace.Wrap(err)
-	} else if isIntegration {
-		return "", trace.NotImplemented("export with secrets is not supported for %q CAs", req.AuthType)
 	}
-	return exportAuth(ctx, client, req, true /* exportSecrets */)
+	// At least one authority is guaranteed on success by both ExportAll methods.
+	if l := len(authorities); l > 1 {
+		return "", trace.BadParameter("export returned %d authorities, expected exactly one", l)
+	}
```

Comment:
Why can't we receive more than 1 authority here?
I understand that `exportAllFunc` will return an error instead of an empty array, but I don't understand where this restriction is coming from. If the function is meant to export a single authority, should it be called `exportAuthority` instead?

Reply:
This one is indeed the single-authority export. I can rename it to exportAuthority instead.

Do keep in mind that this and all the "unary" export functions will be deleted as soon as the follow up PRs land. I'm just keeping them while I do the rest of the changes.

Reply:
Also it's worth saying that it is named after the public funcs that call it: ExportAuthorities and ExportAuthoritiesSecrets (these are the pre-existing ones I'll delete at the end).

Reply:
If you plan to cleanup and have a final state with consistent naming, I don't think it's an issue.